HAL/STM32: Use DMA for Serial driver to reduce IRQ timing requirements and overhead.

gsmcmullin opened the pull request.

gsmcmullin committed:
HAL STM32: Use DMA for Serial RX.

gsmcmullin committed:
HAL STM32: Use DMA for Serial TX.

jacobmcnamee commented on os/hal/ports/STM32/LLD/USARTv1/serial_lld.c:
```diff
@@ -111,9 +168,24 @@ static void usart_init(SerialDriver *sdp, const SerialConfig *config) {
+                               STM32_DMA_CR_CIRC | STM32_DMA_CR_MINC);
+  dmaStreamEnable(sdp->dmarx);
+
+  dmaStreamAllocate(sdp->dmatx, STM32_UART_USART1_IRQ_PRIORITY,
```
generalize

jacobmcnamee commented:
FYI it appears to be impossible to use SPI2 and USART3 with DMA concurrently. USART2 seems to work fine, though.

gsmcmullin committed:
HAL STM32: Fix allocation of Serial DMA IRQ priorities.

gsmcmullin committed:
HAL STM32:  Allow disabling SPI DMA with a configuration option.

Currently trying to use the async SPI routines will fail an assert if DMA is disabled.

gsmcmullin committed:
HAL STM32 Serial: Use IDLE and DMA interrupt instead of RXNE.
